### PR TITLE
Added support for using the "{{ label }}" placeholder in constraint messages

### PR DESCRIPTION
**Pull request by a-menshchikov**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | #12238 
| License       | MIT
| Doc PR        | 

- [x] Add docs PR

### CODE REVIEW

**Comment by a-menshchikov** [comment]
There is no support for `label_format` yet. I have no idea about how get `%id%` substitution value in `ViolationMapper`.

File: src/Symfony/Component/Form/Extension/Validator/ValidatorExtension.php [hunk 1]
```diff
@@ -24,9 +26,22 @@
...
     private $validator;
 
-    public function __construct(ValidatorInterface $validator)
+    /**
+     * @var FormRendererInterface
+     */
+    private $formRenderer;
+
+    /**
+     * @var TranslatorInterface
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Validator/ValidatorExtension.php]: Please remove all added annotation on properties, the constructor is enough to define them.

**Comment by nicolas-grekas** [comment]
I think this would need some examples. Reviewing I'm wondering e.g. why doesn't the message go through translation? Why do we use str_replace instead of "label" as a translation variable? Should the translating the label be done before or after humanize() by form? etc. All these may become more obvious with examples.

**Comment by a-menshchikov** [comment]
> Why doesn't the message go through translation?
Message into violation object already translated.

> Why do we use str_replace instead of "label" as a translation variable?
Since the message has already been translated, it makes no sense to translate it one more time with `{{ label }}` parameter. IMO str_replace is quite enough. (But maybe I'm not taking something into account.)

> Should translating the label be done before or after humanize() by form?
There I've repeated translation logic from `form_div_layout.html.twig` because we want to get the same label as it rendered on form.

**Comment by a-menshchikov** [comment]
Maybe placeholder should be `%label%` for consistence with substitutions format in `label_format` widget option?

**Comment by a-menshchikov** [comment]
@nicolas-grekas what can I do to bring the review closer?

**Comment by fabpot** [comment]
@xabbuh Could you review this PR?

File: src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php [hunk 2]
```diff
@@ -28,10 +30,10 @@ class FormTypeValidatorExtension extends BaseValidatorExtension
     private $violationMapper;
     private $legacyErrorMessages;
 
-    public function __construct(ValidatorInterface $validator, bool $legacyErrorMessages = true)
+    public function __construct(ValidatorInterface $validator, FormRendererInterface $formRenderer = null, TranslatorInterface $translator = null, bool $legacyErrorMessages = true)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php:33]: Changing the position of the `$legacyErrorMessages` is a BC break.

File: src/Symfony/Component/Form/Extension/Validator/ValidatorExtension.php [hunk 3]
```diff
@@ -25,9 +27,11 @@
 class ValidatorExtension extends AbstractExtension
 {
     private $validator;
+    private $formRenderer;
+    private $translator;
     private $legacyErrorMessages;
 
-    public function __construct(ValidatorInterface $validator, bool $legacyErrorMessages = true)
+    public function __construct(ValidatorInterface $validator, FormRendererInterface $formRenderer = null, TranslatorInterface $translator = null, bool $legacyErrorMessages = true)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Validator/ValidatorExtension.php:34]: BC break also

File: src/Symfony/Component/Form/Extension/Validator/ViolationMapper/ViolationMapper.php [hunk 4]
```diff
@@ -13,22 +13,33 @@
...
 /**
  * @author Bernhard Schussek <bschussek@gmail.com>
  */
 class ViolationMapper implements ViolationMapperInterface
 {
+    private $formRenderer;
+    private $translator;
+
     /**
      * @var bool
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Validator/ViolationMapper/ViolationMapper.php:33]: while doing so, let's remove this annotation and make the property default to false

File: src/Symfony/Component/Form/Extension/Validator/ViolationMapper/ViolationMapper.php [hunk 5]
```diff
@@ -124,9 +131,33 @@ public function mapViolation(ConstraintViolation $violation, FormInterface $form
 
         // Only add the error if the form is synchronized
         if ($this->acceptsErrors($scope)) {
+            $label = $scope->getConfig()->getOption('label');
+
+            if (null === $label && null !== $this->formRenderer) {
```
Comment (xabbuh) [inline src/Symfony/Component/Form/Extension/Validator/ViolationMapper/ViolationMapper.php:136]: I think we first need to take into account the `label_format` option.

**Review by xabbuh — COMMENTED** [review]
Just one minor comment. Apart from that the PR looks very good. 👍

**Review by xabbuh — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @a-menshchikov.

**Comment by fabpot** [comment]
@a-menshchikov Can you submit a PR for the docs?

**Pull request merged**